### ISSUE TITLE
Error building static libraries on Ubuntu 15.10

### ISSUE THREAD

**Issue by rmcgibbo** [Unidata/netcdf-c#174]
I see an error building the current master configured with `cmake -DBUILD_SHARED_LIBS=OFF` on Ubuntu 15.10, using the system package manager's `libhdf5-dev` package for hdf5 and `gcc (Ubuntu 5.2.1-22ubuntu2) 5.2.1 20151010` throws the following error.

```
make VERBOSE=1
[...]
Linking C executable ncgen
cd /home/rmcgibbo/projects/netcdf-c/build/ncgen && /usr/bin/cmake -E cmake_link_script CMakeFiles/ncgen.dir/link.txt --verbose=1
/usr/bin/cc  -g -Wall -Wconversion   CMakeFiles/ncgen.dir/generate.c.o CMakeFiles/ncgen.dir/main.c.o CMakeFiles/ncgen.dir/cdata.c.o CMakeFiles/ncgen.dir/bindata.c.o CMakeFiles/ncgen.dir/genchar.c.o CMakeFiles/ncgen.dir/cvt.c.o CMakeFiles/ncgen.dir/data.c.o CMakeFiles/ncgen.dir/debug.c.o CMakeFiles/ncgen.dir/escapes.c.o CMakeFiles/ncgen.dir/genc.c.o CMakeFiles/ncgen.dir/genbin.c.o CMakeFiles/ncgen.dir/generr.c.o CMakeFiles/ncgen.dir/genlib.c.o CMakeFiles/ncgen.dir/getfill.c.o CMakeFiles/ncgen.dir/odom.c.o CMakeFiles/ncgen.dir/offsets.c.o CMakeFiles/ncgen.dir/semantics.c.o CMakeFiles/ncgen.dir/ncgeny.c.o CMakeFiles/ncgen.dir/dump.c.o CMakeFiles/ncgen.dir/util.c.o CMakeFiles/ncgen.dir/bytebuffer.c.o CMakeFiles/ncgen.dir/list.c.o CMakeFiles/ncgen.dir/genf77.c.o CMakeFiles/ncgen.dir/f77data.c.o CMakeFiles/ncgen.dir/genj.c.o CMakeFiles/ncgen.dir/jdata.c.o CMakeFiles/ncgen.dir/nc_iter.c.o CMakeFiles/ncgen.dir/ConvertUTF.c.o  -o ncgen -rdynamic ../liblib/libnetcdf.a /usr/lib/x86_64-linux-gnu/hdf5/serial/lib/libhdf5_hl.a /usr/lib/x86_64-linux-gnu/hdf5/serial/lib/libhdf5.a -Wl,-Bstatic -lpthread -lz -ldl -lm -Wl,-Bdynamic -ldl -lm -lz -lcurl -Wl,-rpath,::::::::::::::::::::::
/usr/lib/x86_64-linux-gnu/hdf5/serial/lib/libhdf5.a(H5PL.o): In function `H5PL__open':
/build/hdf5-LYDWMG/hdf5-1.8.15-patch1+docs/debian/build/src/../../../src/H5PL.c:596: warning: Using 'dlopen' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
/usr/lib/gcc/x86_64-linux-gnu/5/../../../x86_64-linux-gnu/libpthread.a(pthread_create.o): In function `allocate_stack':
/build/buildd/glibc-2.21/nptl/allocatestack.c:475: undefined reference to `_dl_stack_flags'
/build/buildd/glibc-2.21/nptl/allocatestack.c:613: undefined reference to `_dl_stack_flags'
/usr/lib/gcc/x86_64-linux-gnu/5/../../../x86_64-linux-gnu/libpthread.a(nptl-init.o): In function `__pthread_initialize_minimal_internal':
/build/buildd/glibc-2.21/nptl/nptl-init.c:303: undefined reference to `__libc_setup_tls'
/build/buildd/glibc-2.21/nptl/nptl-init.c:323: undefined reference to `_dl_cpuclock_offset'
/build/buildd/glibc-2.21/nptl/nptl-init.c:450: undefined reference to `_dl_pagesize'
/build/buildd/glibc-2.21/nptl/nptl-init.c:459: undefined reference to `_dl_pagesize'
/build/buildd/glibc-2.21/nptl/nptl-init.c:479: undefined reference to `_dl_init_static_tls'
/build/buildd/glibc-2.21/nptl/nptl-init.c:481: undefined reference to `_dl_wait_lookup_done'
/usr/lib/gcc/x86_64-linux-gnu/5/../../../x86_64-linux-gnu/libpthread.a(nptl-init.o): In function `__pthread_get_minstack':
/build/buildd/glibc-2.21/nptl/nptl-init.c:502: undefined reference to `_dl_pagesize'
/usr/lib/gcc/x86_64-linux-gnu/5/../../../x86_64-linux-gnu/libdl.a(dlopen.o): In function `dlopen':
(.text+0x5): undefined reference to `__dlopen'
/usr/lib/gcc/x86_64-linux-gnu/5/../../../x86_64-linux-gnu/libdl.a(dlclose.o): In function `dlclose':
(.text+0x1): undefined reference to `__dlclose'
/usr/lib/gcc/x86_64-linux-gnu/5/../../../x86_64-linux-gnu/libdl.a(dlsym.o): In function `dlsym':
(.text+0x5): undefined reference to `__dlsym'
/usr/lib/gcc/x86_64-linux-gnu/5/../../../x86_64-linux-gnu/libdl.a(dlerror.o): In function `dlerror':
(.text+0x1): undefined reference to `__dlerror'
/usr/lib/gcc/x86_64-linux-gnu/5/../../../x86_64-linux-gnu/libpthread.a(ptw-write.o): In function `__write_nocancel':
/build/buildd/glibc-2.21/nptl/../sysdeps/unix/syscall-template.S:81: undefined reference to `__syscall_error'
/build/buildd/glibc-2.21/nptl/../sysdeps/unix/syscall-template.S:81: undefined reference to `__syscall_error'
/usr/lib/gcc/x86_64-linux-gnu/5/../../../x86_64-linux-gnu/libpthread.a(ptw-read.o): In function `__read_nocancel':
/build/buildd/glibc-2.21/nptl/../sysdeps/unix/syscall-template.S:81: undefined reference to `__syscall_error'
/build/buildd/glibc-2.21/nptl/../sysdeps/unix/syscall-template.S:81: undefined reference to `__syscall_error'
/usr/lib/gcc/x86_64-linux-gnu/5/../../../x86_64-linux-gnu/libpthread.a(ptw-close.o): In function `__close_nocancel':
/build/buildd/glibc-2.21/nptl/../sysdeps/unix/syscall-template.S:81: undefined reference to `__syscall_error'
/usr/lib/gcc/x86_64-linux-gnu/5/../../../x86_64-linux-gnu/libpthread.a(ptw-close.o):/build/buildd/glibc-2.21/nptl/../sysdeps/unix/syscall-template.S:81: more undefined references to `__syscall_error' follow
collect2: error: ld returned 1 exit status
ncgen/CMakeFiles/ncgen.dir/build.make:771: recipe for target 'ncgen/ncgen' failed
make[2]: *** [ncgen/ncgen] Error 1
make[2]: Leaving directory '/home/rmcgibbo/projects/netcdf-c/build'
CMakeFiles/Makefile2:1486: recipe for target 'ncgen/CMakeFiles/ncgen.dir/all' failed
make[1]: *** [ncgen/CMakeFiles/ncgen.dir/all] Error 2
make[1]: Leaving directory '/home/rmcgibbo/projects/netcdf-c/build'
Makefile:146: recipe for target 'all' failed
make: *** [all] Error 2
```

If you squint closely, the linker invocation it is `-rdynamic ../liblib/libnetcdf.a /usr/lib/x86_64-linux-gnu/hdf5/serial/lib/libhdf5_hl.a /usr/lib/x86_64-linux-gnu/hdf5/serial/lib/libhdf5.a -Wl,-Bstatic -lpthread -lz -ldl -lm -Wl,-Bdynamic -ldl -lm -lz -lcurl -Wl,-rpath,::::::::::::::::::::::`. Note that `-ldl` appears twice. If you remove `-ldl` from the static section and move `-lpthread` into the dynamic section, then it works.

Here's the configuration:
```
# NetCDF C Configuration Summary
==============================

# General
-------
NetCDF Version:		4.4.0-development
Configured On:		Tue Dec  8 11:52:02 PST 2015
Host System:		x86_64-Linux-4.2.0-18-generic
Build Directory: 	/home/rmcgibbo/projects/netcdf-c/build
Install Prefix:         /home/rmcgibbo/usr

# Compiling Options
-----------------
C Compiler:		/usr/bin/cc
CFLAGS:			 -g -Wall -Wconversion
CPPFLAGS:
LDFLAGS:		  -Wl,--no-undefined
AM_CFLAGS:
AM_CPPFLAGS:
AM_LDFLAGS:
Shared Library:		no
Static Library:		yes
Extra libraries:	-lhdf5_hl -lhdf5 -lpthread -lz -ldl -lm -ldl -lm -lz -lcurl

# Features
--------
NetCDF-2 API:		yes
HDF4 Support:		no
NetCDF-4 API:		yes
NC-4 Parallel Support:	no
PNetCDF Support:	no
DAP Support:		yes
Diskless Support:	yes
MMap Support:		no
JNA Support:		no
```

**Comment by WardF** [comment]
I am unable to duplicate this issue under Ubuntu and as described.  Can you provide a pull request making the changes you suggest?

**Comment by rmcgibbo** [comment]
I'm actually not sure how to fix it.

**Comment by rmcgibbo** [comment]
Try the following to reproduce

```
# docker run -i -t ubuntu:wily /bin/bash
```

Then run:
```
# apt-get update && apt-get install git-all cmake build-essential gcc g++ libhdf5-dev m4 libcurl4-openssl-dev
# git clone https://github.com/Unidata/netcdf-c.git
# cd netcdf-c && mkdir build && cd build
# cmake -DBUILD_SHARED_LIBS=OFF -DCMAKE_BUILD_TYPE=Release ..
# make
```
(https://gist.github.com/rmcgibbo/9e8782f4a4dded821d21)

**Comment by WardF** [comment]
It looks like this will be non-trivial to build.  I've cleaned up our cmake files a bit, but right now I think the solution will be to either build using autotools (which works), or build with cmake but use shared libraries.  Alternatively, you should be able to build statically if you install hdf5 from source and disable pthreads.  This is how our test system is configured.

**Comment by rmcgibbo** [comment]
> It looks like this will be non-trivial to build.

I don't quite understand. My point in using docker is to emphasize that this can be reproduced with a standard "stock" ubuntu 15.10 installation. Docker is not required -- any Ubuntu installation will do.

**Comment by WardF** [comment]
My apologies for being unclear; it was late, and I most likely meant to type "It looks like this will be non-trivial to fix".   I was able to handily recreate the issue in docker, thanks for the suggestion.  I was using my own docker image to try to recreate the issue initially, but that contains hand-built hdf5 libraries.  The issue seems to lie between cmake and hdf5.

The reason it's flagged can't fix is due more to the resource-constrained nature of things right now, combined with the fact that there are work-arounds on Ubuntu 15.10 which do work with the pre-compiled hdf5 libraries.  I tagged it revisit, as this is something I *do* want to fix, and it should be fixable; I spent a couple hours on it last night and I just don't have the time before AGU next week to put any more time into it in the short term.
